compiler/next/include/chpl/uast/For.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_UAST_FOR_H
+#define CHPL_UAST_FOR_H
+
+#include "chpl/queries/Location.h"
+#include "chpl/uast/Loop.h"
+
+namespace chpl {
+namespace uast {
+
+
+/**
+  This class represents a for loop. For example:
+
+  \rst
+  .. code-block:: chapel
+
+      // Example 1:
+      for i in foo {
+        var x;
+      }
+
+  \endrst
+
+ */
+class For final : public Loop {
+ private:
+  For(ASTList children, int8_t indexVarChildNum, int8_t iterandChildNum,
+      bool usesDo,
+      bool expressionLevel,
+      bool param);
+  bool contentsMatchInner(const ASTNode* other) const;
+  void markUniqueStringsInner(Context* context) const;
+  int8_t indexVarChildNum_;
+  int8_t iterandChildNum_;
+  bool expressionLevel_;
+  bool param_;
+
+ public:
+
+  /**
+    Create and return a for loop. 
+  */
+  static owned<For> build(Builder* builder, Location loc,
+                          owned<Expression> indexVar,
+                          owned<Expression> iterand,
+                          ASTList stmts,
+                          bool usesDo,
+                          bool expressionLevel,
+                          bool param);
+
+  /**
+    Create and return a for loop without an index variable.
+  */
+  static owned<For> build(Builder* builder, Location loc,
+                          owned<Expression> iterand,
+                          ASTList stmts,
+                          bool usesDo);
+
+  /**
+    Returns the index variable of this for loop.
+  */
+  const Expression* indexVar() const {
```

Comment:
It looks like you are currently representing the `indexVar` as an `Identifier` but it's actually a new variable, so I would expect that the loop contain a `VariableDecl` for it.

Then, this function should probably have two variants:
 * `indexVariableDecl` returns a `VariableDecl*`
 * `indexVariable` returns a `Variable*`
 
I don't currently know of a reason why the VariableDecl would be needed but it might be and it's easy enough to include. It will not be easy to go from the indexVariable back to the Decl (since uast does not have links to the parent).

Reply:
I had a side conversation with @lydia-duncan about this and the conclusion of that conversation is:
 * the normal case is wanting to get to the `Sym` for the declared variable (so `indexVariable` should return `Variable*`)
 * we would like to be able to access the `Decl` in case it is needed, so would want the 2nd variant (e.g. `indexVariableDecl` above). This is important because the uAST nodes don't have links back to parent nodes.